client/src/components/ElementEditor/ElementList.js

```diff
@@ -11,6 +11,101 @@ import { getDragIndicatorIndex } from 'lib/dragHelpers';
 import { getElementTypeConfig } from 'state/editor/elementConfig';
 
 class ElementList extends Component {
+  constructor(props) {
+    super(props);
+    this.resetState = this.resetState.bind(this);
+    this.handleBeforeSubmitForm = this.handleBeforeSubmitForm.bind(this);
+    this.handleAfterSubmitResponse = this.handleAfterSubmitResponse.bind(this);
+    this.state = {
+      // saveAllElements will be set to true in entwine.js in the 'onbeforesubmitform' "hook"
+      // which is triggered by LeftAndMain submitForm()
+      saveAllElements: false,
+      // increment is also set in entwine.js in the 'onbeforesubmitform' "hook"
+      increment: 0,
+      hasUnsavedChangesBlockIDs: {},
+      validBlockIDs: {},
+    };
+    // Update the sharedObject so that setState() can be called from entwine.js
+    this.props.sharedObject.setState = this.setState.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // Number of blocks just changed after graphql response
+    if (this.props.blocks !== prevProps.blocks) {
+      this.resetState(prevState, false);
+      return;
+    }
```

Comment:
Can you please reword this comment? I'm not sure what you mean by it. Is it something to do with adding new or removing existing blocks?
Does the sort order matter?
Why are we resetting the state anyway?
Why do we not do all the rest of the logic below?

Reply:
Updated

Reply:
What was changed? I still don't know the answers to any of the questions I asked.

Reply:
@emteknetnz Please respond to this

Reply:
Comment was changed to `// Scenario: Number of blocks just changed after graphql response`

For the other questions .. I don't know / can't remember - It's been weeks since I wrote this.  I guess test it and if it doesn't behave as you'd expect it to then we could change osmething.

Reply:
The original comment: "Number of blocks just changed after graphql response"
The new comment: "Scenario: Number of blocks just changed after graphql response"

All you did was put the word "Scenario" in front of it. It still doesn't explain what's happening sufficiently, which is why I have all of those questions - and why you now aren't able to answer them.

I don't know what this is doing because the comment doesn't explain it, so I don't know how to craft a test that will specifically test for the questions I listed above.

Reply:
> Is it something to do with adding new or removing existing blocks?

It's in the componentDidUpdate() method, and we're checking that `(this.props.blocks !== prevProps.blocks) ` means that this.props.blocks changed.

this.props.blocks comes from graphql ... somehow (the graphql implementation is so confusing I actually don't know how it populates the prop)

Hence the comment `"Scenario: Number of blocks just changed after graphql response".  I'll update the comment to say "Scenario: blocks props just changed after a graphql query response updated it" because that's more accurate.

> Does the sort order matter?

I'm not sure quite what this means, though if you drag and drap the elements this it will trigger this code block. There is a graphql request made whenever you drag and drop sort elements.

> Why are we resetting the state anyway?

If we've retched the blocks then we've essentially reset the component, so the state needs to be reset

> Why do we not do all the rest of the logic below?

Because it's for a totally different scenario `// Scenario Saving all elements and state has just updated because of a formSchema response from an inline save - see Element.js handleFormSchemaSubmitResponse()`

This is in the context of the `componentDidUpdate()` method which will trigger in any change in state or props. This is a class component, if it were a functional component I'd be wrapping everything in their own `useEffect()` which would make it more obvious that they're totally separate things